pkg/service/did/ion.go

```diff
@@ -88,6 +91,287 @@ func (i ionStoredDID) IsSoftDeleted() bool {
 	return i.SoftDeleted
 }
 
+type UpdateDIDRequest struct {
+	DID ion.ION `json:"did"`
+
+	StateChange ion.StateChange `json:"stateChange"`
+}
+
+type UpdateDIDResponse struct {
+	DID did.Document `json:"did"`
+}
+
+type preAnchor struct {
+	updateOperation      *ion.UpdateRequest
+	nextUpdatePublicJWK  *jwx.PublicKeyJWK
+	nextUpdatePrivateJWK *jwx.PrivateKeyJWK
+	storedDID            *ionStoredDID
+}
+
+type anchor struct {
+	// The result of calling anchor.
+	err error
+}
+
+type updateState struct {
+	status    UpdateRequestStatus
+	preAnchor *preAnchor
+	anchor    *anchor
+}
+
+func (h *ionHandler) UpdateDID(ctx context.Context, request UpdateDIDRequest) (*UpdateDIDResponse, error) {
+	updateStates, updatePrivateKey, err := h.readUpdateState(ctx, request.DID.String())
+	if err != nil {
+		return nil, err
+	}
+	state := &updateStates[len(updateStates)-1]
+	if state.status == DoneStatus {
+		updateStates = append(updateStates, updateState{})
+		state = &updateStates[len(updateStates)-1]
+	}
+
+	watchKeys := []storage.WatchKey{
+		{
+			Namespace: updateRequestStatusNamespace,
+			Key:       request.DID.String(),
+		},
+	}
+
+	if state.status == "" {
+		execResp, err := h.storage.db.Execute(ctx, func(ctx context.Context, tx storage.Tx) (any, error) {
+			didSuffix, err := request.DID.Suffix()
+			if err != nil {
+				return nil, errors.Wrap(err, "getting did suffix")
+			}
+
+			updateKey := updatePrivateKey.ToPublicKeyJWK()
+
+			signer, err := ion.NewBTCSignerVerifier(*updatePrivateKey)
+			if err != nil {
+				return nil, errors.Wrap(err, "creating btc signer verifier")
+			}
+
+			nextUpdateKey, nextUpdatePrivateKey, err := h.nextUpdateKey()
+			if err != nil {
+				return nil, err
+			}
+
+			updateOp, err := ion.NewUpdateRequest(didSuffix, updateKey, *nextUpdateKey, *signer, request.StateChange)
+			if err != nil {
+				return nil, errors.Wrap(err, "creating update request")
+			}
+
+			state.preAnchor = &preAnchor{
```

Comment:
this should be blocked until 539 is in, no?

Reply:
Depends on how much pushback there will be for #539 :)